### ISSUE TITLE
`jsdoc/valid-types` Error on valid type signature in typescript mode.

### ISSUE THREAD

**Issue by PatrickHollweck** [gajus/eslint-plugin-jsdoc#646]
## Expected behavior

To parse the typescript type declaration

## Actual behavior

Emits a `jsdoc/valid-types` warning.

## ESLint Config

```js
module.exports = {
  root: true,
  extends: [
    "eslint:recommended",
    "plugin:jsdoc/recommended",
  ],
  parser: "babel-eslint",
  parserOptions: {
    ecmaVersion: 9,
    sourceType: "module",
  },
  env: {
    es6: true,
    node: true,
    jquery: true,
    browser: true,
  },
  plugins: ["jsdoc"],
  rules: {
    "jsdoc/require-jsdoc": [0],
    "jsdoc/require-param-description": [0],
    "jsdoc/require-returns-description": [0],
  },
  settings: {
    jsdoc: {
        mode: "typescript"
    }
  }
};
```

## ESLint sample

Both the @param and the @returns currently trigger a warning.
Description: `Syntax error in type: { dispatch: (action: object) => void}`.

I don't know if this is related to this package or the parser. But I suspect the config is not applied.
However I can use most other typescript types declarations: `["A", "B"]` or `any`.
The parser package states that it does support these (`() => void` and `import(...)`) syntaxes.

```js
    /**
     * Detect and setup the best matching display mode.
     *
     * @param {{ dispatch: (action: object) => void }} store The store
     * @returns {import('react').ReactElement} The element to render
     */
    static setupDisplayMode(store) { /* ... */ }
```

## Environment

- Node version: v12.18.2
- ESLint version: v7.10.0
- `eslint-plugin-jsdoc` version: v30.6.3

**Comment by brettz9** [comment]
I don't know if the more modern version of `babel-eslint`, [@babel/eslint-parser](https://babeljs.io/blog/2020/07/13/the-state-of-babel-eslint), supports TypeScript, but I have confirmed this is not an issue with `@typescript-eslint/parser` as the parser. You should first see whether your parser supports or should support TypeScript.

**Comment by PatrickHollweck** [comment]
Hey, I migrated to @babel/eslint-parser and this did not fix the issue.
I dont know in how far the eslint-parser is related to the parsing of the doc blocks.

Is there anything I can do?

**Comment by brettz9** [comment]
Is there a reason you can't use `@typescript-eslint/parser`?

**Comment by PatrickHollweck** [comment]
Well, my companies codebase is (sadly) written in javascript, with some babel plugins.

I don't know if @typescript-eslint is compatible with my setup. But that is nothing you have to concern yourself with...
Now that I think about it, since typescript is a superset of javascript it should work. Have to test the eslint plugins though.

Simply put, If you tell me that it will work with the @typescript-eslint parser that I will migrate to that.
This may as well be the "solution" to this issue. Feel free to close

**Comment by brettz9** [comment]
I can't make any guarantees, and `@typescript-eslint/parser` might not support some experimental features of JavaScript (see https://github.com/typescript-eslint/typescript-eslint#what-about-babel-and-babel-eslint ).

However, since TS is indeed a superset and that parser supports TS up to 4.0.*, and TS 3.8 added ES2020 support, that parser should work for even the latest stable JavaScript. A lot of our users are using it. You'd have to do your own investigations, but it does seem to me to be a safe bet if your project is not using `babel-eslint` to take advantage of experimental JS features (note that some formerly experimental features, like `import()` have already become a part of ES ES2020).

**Comment by PatrickHollweck** [comment]
Yeah, we dont use any crazy babel plugins. Only some transforms for compatibility with older browsers.
I will do some experimentation and see how it goes.

Thanks for your efforts :) <3
